[ADT] Refactor StringMap iterators (NFC)

kazutakahirata opened the pull request:
StringMap has four iterator classes:

- StringMapIterBase
- StringMapIterator
- StringMapConstIterator
- StringMapKeyIterator

This patch consolidates the first three into one class, namely
StringMapIterBase, adds a boolean template parameter to indicate
desired constness, and then use "using" directives to specialize the
common class:

  using const_iterator = StringMapIterBase<ValueTy, true>;
  using iterator = StringMapIterBase<ValueTy, false>;

just like how we simplified DenseMapIterator.

Remarks:

- This patch drops CRTP and iterator_facade_base for simplicity.  For
  fairly simple forward iterators, iterator_facade_base doesn't buy us
  much.  We just have to write a few "using" directives and operator!=
  manually.

- StringMapIterBase has a SFINAE-based constructor to construct a
  const iterator from a non-const one just like DenseMapIterator.

- We now rely on compiler-generated copy and assignment operators.

kazutakahirata committed:
[ADT] Refactor StringMap iterators (NFC)

StringMap has four iterator classes:

- StringMapIterBase
- StringMapIterator
- StringMapConstIterator
- StringMapKeyIterator

This patch consolidates the first three into one class, namely
StringMapIterBase, adds a boolean template parameter to indicate
desired constness, and then use "using" directives to specialize the
common class:

  using const_iterator = StringMapIterBase<ValueTy, true>;
  using iterator = StringMapIterBase<ValueTy, false>;

just like how we simplified DenseMapIterator.

Remarks:

- This patch drops CRTP and iterator_facade_base for simplicity.  For
  fairly simple forward iterators, iterator_facade_base doesn't buy us
  much.  We just have to write a few "using" directives and operator!=
  manually.

- StringMapIterBase has a SFINAE-based constructor to construct a
  const iterator from a non-const one just like DenseMapIterator.

- We now rely on compiler-generated copy and assignment operators.

kuhar reviewed (commented):
Do we have dedicated unit tests for these iterators?

kuhar commented:
Maybe we could add some tests with `is_same_v` to check that the types are as expected, e.g., that dereferencing const iterator give you the right const reference etc?

kazutakahirata committed:
Address a comment.

kazutakahirata committed:
Make value_type a non-const StringMapEntry<ValueTy>.

kazutakahirata commented:
@kuhar I've added a few more tests on iterator traits.  Also, I've corrected a bug where `value_type` was a const value type in `const_iterator`.  IIUC, `value_type` is always a non-const value type.

kazutakahirata committed:
Fix formatting.
